_posts/2018-05-07-is-graphql-the-future.markdown

```diff
@@ -0,0 +1,196 @@
+
+For brevity, the following assumes a intermediate familiarity with GraphQL, including its type system, syntax, and server-side implementation. If you don’t have this level of familiarity, I recommend going through any tutorial that requires you to set up a GraphQL server, not just play with the query language (which is how I ended up with a lot of misconceptions). I’m going to start with the basics, but only so I can put my own spin on those concepts, not to really illustrate them with examples.
+
+# A tree of fetches
+
+Most applications are designed in the form of discrete pages, which are seeded with some tiny chunk of data—say, a key or slug for some domain object—and then perform a cascade of contingent fetches to get the data needed to populate the templates rendered to a user. This is the basis of designing applications driven by URL-based routing and it has been a mainstay of the MVC approach to web application architecture for the past decade.
+
+> **Example:** At Artsy, the seed of data for rendering an artwork page could be the slug identifying some artwork. From this slug, we need a whole bunch more data: the metadata of the artwork, information about the artist(s), sales data if it’s available for purchase, information about the Artsy partner that owns it, and so on. In classic REST, this data is aggregated by a cascade of dozens of HTTP fetches to our backend API.
+
+I wasn’t in the room when GraphQL was invented, but it seems to me that the team that built it made a particularly crucial insight:
+
+> In most cases, all of this contingent fetching forms a tree, which is more or less _fixed_ for a given page.
+
+Data from early responses contain the keys for subsequent requests, but the linkages between these requests are usually straightforward. So if it were possible to factor all this disparate fetching into one spot and encode it into one big “fetching tree” data structure ahead of time, this tree could be sent to the the server, and the server could fulfill all of the data requirements in one shot. This cuts out a tremendous amount of wasteful chatter between client and server. Even in today's broadband world, bandwidth and latency matter, especially for mobile users.
+
+# GraphQL anatomy
+
+A GraphQL always request starts with at least _one root API operation_ and some finite number of follow-up operations. Idiomatically, these follow-up operations are queries, meaning that they just return data, without changing the server state in observable ways. Every operation—-root or otherwise—-can specify some finite number of follow-up operations. GraphQL models API operations as **fields**. How a field works in GraphQL depends on its **type**, which falls into one of two basic categories:
```

Comment:
"A GraphQL request always..."

Reply:
> Every operation—-root or otherwise—-can specify some finite number of follow-up operations.

I feel like this could be lightly expanded in-line with a metaphor to make more concrete -- "...some finite number of follow-up operations, as if..."